src/polydatum/dal.py

```diff
@@ -50,9 +266,16 @@ def _init_service(self, key, service):
         self._services[key] = service
         return service
 
-    def __getattr__(self, name):
-        assert self._data_manager.get_active_context(), 'A DataAccessContext must be started to access the DAL.'
-        return self._services[name]
+    def _call(self, path: Tuple[PathSegment, ...], *args, **kwargs):
+        return self._handler(
+            request=DalMethodRequest(
+                self._data_manager.require_active_context(), path, args, kwargs
```

Comment:
There's a neat feature here. You can do:

```python
# No context yet
getter = dm.get_dal().some_service.example_service.get
# Make a context
with dm.context():
    getter()
```

The context isn't actually required until you try to call the DAL method. This might be useful for passing around DAL methods as hooks or handlers at startup. It might also help with monkypatching in tests.

Reply:
With a creeping horror, I'm hating this feature more and more... It looks like it would enable some really tricky bugs to track down... But I don't think we can stop that. Also, I'm not everyone's nanny, so maybe I just need to get over it :P

Reply:
Actually, I see your other comment, and I feel better. I thought we did this a bunch in the tests, but we test the DalMethodRequest objects directly, so we make things that look like dal service access stuff, but aren't actually being done directly on the dal, specifically to avoid that assertion error (we were testing the functionality directly, didn't need a real context, so we avoided it).

Reply:
I think being able to make `DalMethodRequestors` without a context is great. It allows you to reference other methods at startup time. This is especially important for registering event handlers. You can't register an event handler with a function bound to a specific context.